### ISSUE TITLE
Issue with SOLWEIG processing tool and anisotropic sky

### ISSUE THREAD

**Issue by bweeding** [UMEP-dev/UMEP#204]
When trying to run SOLWEIG using the processing tool, I'm encountering an error when using the anisotropic sky option. The same settings work perfectly well when I using the drop down menu in the GUI. I've included the run info and error printout below. Any advice would be welcomed.

Cheers
Ben

![image](https://user-images.githubusercontent.com/57697604/102032741-ba897680-3e0d-11eb-8f24-6732ff1ca2a8.png)

**Comment by biglimp** [comment]
I cannot replicate your error using my data. Could you please attach all your input data so that I can try to replicate your issue. Could you also please state what version of QGIS and processing for UMEP you are using as well as what OS you are using.

**Comment by biglimp** [comment]
One thing that could be related to your issue is that you have runned SVF calculator without vegetation.

**Comment by bweeding** [comment]
I've attached my DEM, DSM, and CDSM. I double checked this morning and made sure I used vegetation when generating my SVF and still had the same error. 

I'm using the 0.5 version of UMEP for processing, and QGIS 3.16 on Windows 10.


[LAS_fusion_processing.zip](https://github.com/UMEP-dev/UMEP/files/5691870/LAS_fusion_processing.zip)

**Comment by biglimp** [comment]
Could tou please also attach your meteorological file.

**Comment by bweeding** [comment]
Ah, my apologies!

I've used this metfile many other times with the gui and processing tools.

[metfile_20190105_20190108.txt](https://github.com/UMEP-dev/UMEP/files/5694626/metfile_20190105_20190108.txt)

**Comment by biglimp** [comment]
Your metfile looks strange. Your Kdown have positive values during night. Maybe it is in UTC=0. Adjust it to be local time and change UTC in the model interface to the location where your DSM is located accordingly.

**Comment by bweeding** [comment]
My metfile is in UTC 0 and has worked fine in the past... I've just set the
UTC offset to zero.

**Comment by biglimp** [comment]
Ok, I will check. I directly see that your DSM and CDSM is not correct. The DSM should be meter above sea level and include both ground and building elevation. See the tutorial data for SOLWEIG or SEBE. Your CDSM also included some building pixels that should be removed.

**Comment by bweeding** [comment]
Hmmm I produced the DSM and CDSM following the tutorial - perhaps I've made a mistake. I have attached the code I used to produced them. Sorry I haven't yet commented it. I know I need to adjust the CDSM to filter out some building points.
[DSM_CDSM_code.txt](https://github.com/UMEP-dev/UMEP/files/5695349/DSM_CDSM_code.txt)

**Comment by biglimp** [comment]
I cannot replicate your error. I used the SVFCalculator and Wall height and Aspect tools from the GUI then executed SOLWEIG using the processing tool. I recommend to re-run the SVFCalculator and try again.

![image](https://user-images.githubusercontent.com/11908356/102207857-f2a4be00-3ece-11eb-852b-1276db2e8e75.png)

Regarding your DSM and CDSM, a quick fix would be just to add the DEM-values to the DSM to derive a correct DSM. Then run SOLWEIG from GUI saving a building grid and us that to remove vegetation pixel located on buildings on in your CDSM 

I am using QGIS 3.16, W10, UMEP for processing 0.5

You are also correct regarding UTC=0 and your metfile. You can do it like that. I usually use LST but that is not necessary.

**Comment by bweeding** [comment]
Ok great! I'll work on it tomorrow (I'm in Australia) and get back to you.
Thanks for all your help!!

**Comment by biglimp** [comment]
Ok. Please let me know the progress.

**Comment by biglimp** [comment]
Just had a quick look at your code. When you do the DSM, try to not use a ground dtm for normalisation.

```
# DSM
alg_params_dsm = {
    'ADVANCED_MODIFIERS': '',
    'ASCII': True,
    'CELLSIZE': 1,
    'CLASS': '',
    'GROUND': ",
    'INPUT': home_folder+'\\ground_outside_b.las'+';'+home_folder+'\\buildings_inside_b.las',
    'MEDIAN': '',
    'SLOPE': False,
    'SMOOTH': '',
    'VERSION64': True,
    'XYUNITS': 0,
    'ZUNITS': 0,
    'OUTPUT': home_folder+'\\DSM.asc'
```

I like you code. If possible it would be nice to include this in a tutorial for other users. Its up to you and also the data you use needs to be permitted to be distributed to others without and restrictions. Think about it...

**Comment by bweeding** [comment]
Ah ok, perhaps I have followed the instructions incorrectly? I'll try to
get my code tidied up and commented, but it would be great to help with a
tutorial! I aim to get the code working in a standalone python environment
to go from lidar data and met files right through to SOLWEIG output.

I think I'm a bit confused by what the canopy model code does, does it just
take the highest point in each cell and set that as the raster value?

**Comment by biglimp** [comment]
Yes, the highest return is used.

**Comment by bweeding** [comment]
Ah great. I've run into trouble getting the fusion processing tools to load
in the standalone environment. I've tried:

from processing_fusion.fusionProvider import FusionProvider
fusion_provider = FusionProvider()
QgsApplication.processingRegistry().addProvider(fusion_provider)

following the pattern for UMEP, but when I run:

import processing
from processing.core.Processing import Processing
Processing.initialize()

for alg in QgsApplication.processingRegistry().algorithms():
    print("{}:{} --> {}".format(alg.provider().name(), alg.name(),
alg.displayName()))

No fusion tools show up. How do I need to modify the code?

**Comment by bweeding** [comment]
Ah great. I've run into trouble getting the fusion processing tools to load in the standalone environment. I've tried:

`from processing_fusion.fusionProvider import FusionProvider
fusion_provider = FusionProvider()
QgsApplication.processingRegistry().addProvider(fusion_provider)`

following the pattern for UMEP, but when I run:

`import processing
from processing.core.Processing import Processing
Processing.initialize()`

`for alg in QgsApplication.processingRegistry().algorithms():`
    `print("{}:{} --> {}".format(alg.provider().name(), alg.name(), alg.displayName()))`

no fusion tools show up. How do I need to modify the code?

**Comment by biglimp** [comment]
You need to append the path to the tool, did you do that? Something like:
`sys.path.append(r'C:\Users\ **your_username** \AppData\Roaming\QGIS\QGIS3\profiles\default\python\plugins')`

Also, i am not sure if the fusion tools will turn up using your last for loop since this point to all tools in the core. Do you see umep tools in the list?

**Comment by bweeding** [comment]
Ah sorry, should have included more code! I have appended the path:

# import third party processing plugins
sys.path.append(r'C:\Users\weedingb\AppData\Roaming\QGIS\QGIS3\profiles\default\python\plugins')

from processing_umep.processing_umep_provider import ProcessingUMEPProvider
umep_provider = ProcessingUMEPProvider()
QgsApplication.processingRegistry().addProvider(umep_provider)

# not working yet - posted on github
from processing_fusion.fusionProvider import FusionProvider
fusion_provider = FusionProvider()
QgsApplication.processingRegistry().addProvider(fusion_provider)

If I do the above and then run:

import processing
from processing.core.Processing import Processing
Processing.initialize()

for alg in QgsApplication.processingRegistry().algorithms():
    print("{}:{} --> {}".format(alg.provider().name(), alg.name(),
alg.displayName()))

I see all the UMEP plugins but nothing from fusion!?

**Comment by biglimp** [comment]
Hm, I think I know why. The Fusion plugin was originally created by other people and the structure looks different from the UMEP processing plugin. It has an additional file called fusionProviderPlugin.py. I am not sure how it is related to the fusionProvider.py file. This probably affects the syntax used to access fusion in a script. If you cant solve it we could change the structure of the plugin but that will probably not happen until after the new year...

**Comment by bweeding** [comment]
Ah found a solution!

`from processing_fusion.fusionProvider import FusionProvider
fusion_provider = FusionProvider()
fusion_provider.load()
QgsApplication.processingRegistry().addProvider(fusion_provider)`

When the FusionProvider class is instantiated, the list of algorithms is set to null, so you need to run the load method.

**Comment by biglimp** [comment]
hm, this does not work for me. Are you running the code inside of QGIS (in the Python console) or outside? I am using VSCode. I am wondering if that is why...

**Comment by bweeding** [comment]
I'm running it in Spyder through anaconda, completely external to QGIS.

**Comment by biglimp** [comment]
CAn you send your complete imports etc. THats happen before you access fusion, just to see that I have not missed anything.

**Comment by biglimp** [comment]
Works for me in the QGIS Python Concole but not outside. Can you also send me info on your environment path and python path.

**Comment by bweeding** [comment]
Ok I can do that on Tuesday - I am off camping for the weekend.

**Comment by biglimp** [comment]
Actually, since I am also in charge of the fusion repo I will have a go to change the code to look like umep. That will create consistency for other users as well. I let you know about the progress.

**Comment by bweeding** [comment]
Ok sounds good. I think the issue is that when the fusion class is
initialised it sets the list of algorithms to zero. I think if you put the
load method into init it would fix it?

**Comment by biglimp** [comment]
I managed to change the code according to the same way as processing for UMEP. It works for me now but sometime I get a strange error when I try to install the plugin via Manage and Install Plugin in QGIS: 
`RuntimeError: wrapped C/C++ object of type ProcessingFUSIONProvider has been deleted`

Maybe you can try it out. I attach a zipped version.

The new way to access the code outside is now: 
```
from processing_fusion.processing_fusion_provider import ProcessingFUSIONProvider
fusion_provider = ProcessingFUSIONProvider()
QgsApplication.processingRegistry().addProvider(fusion_provider)
```
[processing_fusion.zip](https://github.com/UMEP-dev/UMEP/files/5716537/processing_fusion.zip)

**Comment by biglimp** [comment]
Now I also found the issue with wrapped C/C++ object. https://github.com/jdugge/BufferByPercentage/pull/14/commits/04269ff6b25433253259294eb480a10e4b353403

Please use the attached version. I will update the fusion repo later.
[processing_fusion.zip](https://github.com/UMEP-dev/UMEP/files/5725238/processing_fusion.zip)

**Comment by bweeding** [comment]
The attached fusion file works for me now without the .load method! Thanks!

I will be on Christmas break until the 4th, when I will tidy up the code
and comment it properly. Looking forward to helping produce a tutorial.

Have a good Christmas!
Ben

**Comment by biglimp** [comment]
Great! 
Merry Christmas. 
I will close this issue now and officially update the fusion plugin in the QGIS plugin repository. I will be back on the 7th.

**Comment by bweeding** [comment]
Hope Christmas and New Year treated you well!

Have you set up and run UMEP/SOLWEIG on servers very often? I'm looking to
run my Python code on a Jupyter notebook on a server, and just wondering if
there are any complications with accessing the QGIS processing tools, or if
it should all be straightforward.

Still very keen to help with a tutorial!

Cheers
Ben

**Comment by biglimp** [comment]
No, I didnt try UMEP on a server but theoretically it should work if you have QGIS and any necessery Python libraries installed plus the UMEP for processing available.

**Comment by bweeding** [comment]
Hi, hope you are well! I'm back working on my standalone scripts, getting
ready to port them onto a server. I have encountered an odd issue which I
thought you might have dealt with before.

I am able to use 3rd party algorithms like UMEP just fine, but have no
access to the 'native' algorithms. Have you encountered this before?
I've attached a summary of the problem that I've also posted on GIS
stackexchange.

[image: image.png]
Cheers
Ben
